rust/connlib/tunnel/src/messages.rs

```diff
@@ -169,11 +169,36 @@ pub struct Interface {
     /// DNS that will be used to query for DNS that aren't within our resource list.
     #[serde(skip_serializing_if = "Vec::is_empty")]
     #[serde(default)]
-    pub upstream_dns: Vec<DnsServer>,
+    pub upstream_dns: Vec<DnsServer>, // TODO: Remove once portal with `upstream_do53` support has been deployed.
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    #[serde(default)]
+    pub upstream_do53: Vec<UpstreamDo53>,
+
     #[serde(default)]
     pub search_domain: Option<DomainName>,
 }
 
+impl Interface {
+    pub fn upstream_do53(&self) -> Vec<IpAddr> {
+        if !self.upstream_do53.is_empty() {
+            return self.upstream_do53.iter().map(|u| u.ip).collect();
+        }
+
+        // Fallback whilst the portal does not send `upstream_do53`.
+        self.upstream_dns
```

Comment:
so, in this case, I would probably have not used an early return, and rather just have `if/else` block ;) because that's easier to reason about for me.

Reply:
How do you extend the function if we'd now have to add another condition? By always favouring early-return, you can just add new conditions to it without having to think much about how it interacts with the existing one, plus, the actual change diff is much smaller.

Refactoring from `if-else` to early return or also adding additional `if-else` checks usually incur a much larger diff. Git is unfortunately pretty bad at recognising code that has only been indented but is otherwise unchanged.

Reply:
I think the much bigger problem I have with `else` is that it is completely implicit & context-dependent when it gets executed. `else` always means something different and the smallest change to the boolean condition can completely change, when the `else` block is executed. Constraining yourself to not using `else` essentially means you need to think about how you can structure your code such that the various branches are independent of each other which I believe makes them easier to understand.